import grpc
import sys
import os
import time
from typing import List, Optional, Dict, Tuple

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'proto')))
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'metadata_server')))

import filesystem_extended_pb2 as fs_pb2
import filesystem_extended_pb2_grpc as fs_grpc
from metadata_client2 import MetadataClient

class BigFSClient:
    """Cliente BigFS-v2 com suporte a chunks, replica√ß√£o e toler√¢ncia a falhas"""
    
    def __init__(self, metadata_server: str = "localhost:50052"):
        self.metadata_server = metadata_server
        self.metadata_client = None
        self.storage_connections = {}  # Cache de conex√µes com n√≥s de armazenamento
        
        # Conectar ao servidor de metadados
        self._connect_to_metadata_server()
    
    def _connect_to_metadata_server(self):
        """Conecta ao servidor de metadados"""
        try:
            self.metadata_client = MetadataClient(self.metadata_server)
            print(f"‚úÖ Conectado ao servidor de metadados: {self.metadata_server}")
        except Exception as e:
            print(f"‚ùå Erro ao conectar com servidor de metadados: {e}")
            self.metadata_client = None
    
    def _get_storage_connection(self, node_info: fs_pb2.NodeInfo) -> Optional[fs_grpc.FileSystemServiceStub]:
        """Obt√©m conex√£o com n√≥ de armazenamento (com cache)"""
        if not node_info:
            return None
        node_address = f"{node_info.endereco}:{node_info.porta}"
        
        if node_address not in self.storage_connections:
            try:
                channel = grpc.insecure_channel(
                    node_address,
                    options=[
                        ("grpc.max_send_message_length", 1024 * 1024 * 1024),
                        ("grpc.max_receive_message_length", 1024 * 1024 * 1024)
                    ]
                )
                stub = fs_grpc.FileSystemServiceStub(channel)
                self.storage_connections[node_address] = (channel, stub)
                print(f"üîó Conectado ao n√≥ de armazenamento: {node_address}")
            except Exception as e:
                print(f"‚ùå Erro ao conectar com n√≥ {node_address}: {e}")
                return None
        
        return self.storage_connections[node_address][1]
    
    def verify_file_integrity(self, local_path: str, remote_path: str) -> bool:
        """
        Verifica a integridade de um arquivo completo comparando checksums.
        Primeiro baixa o arquivo, depois compara o checksum com o registrado no servidor.
        """
        if not self.metadata_client:
            print("‚ùå Servidor de metadados n√£o dispon√≠vel")
            return False
        
        try:
            print(f"üîç Iniciando verifica√ß√£o de integridade: {remote_path}")
            
            # 1. Obter metadados do arquivo do servidor
            file_metadata = self.metadata_client.get_file_metadata(remote_path)
            if not file_metadata:
                print(f"‚ùå Arquivo '{remote_path}' n√£o encontrado no servidor")
                return False
            
            expected_checksum = file_metadata.checksum_arquivo
            print(f"üìã Checksum esperado: {expected_checksum}")
            
            # 2. Baixar o arquivo usando a l√≥gica de download paralelo
            success = self.download_file_parallel(remote_path, local_path)
            if not success:
                print(f"‚ùå Falha ao baixar arquivo para verifica√ß√£o")
                return False
            
            # 3. Calcular checksum do arquivo baixado
            import hashlib
            with open(local_path, 'rb') as f:
                file_data = f.read()
                calculated_checksum = hashlib.md5(file_data).hexdigest()
            
            print(f"üìã Checksum calculado: {calculated_checksum}")
            
            # 4. Comparar checksums
            if calculated_checksum == expected_checksum:
                print(f"‚úÖ Integridade do arquivo '{remote_path}' confirmada!")
                return True
            else:
                print(f"‚ùå Integridade comprometida! Checksums n√£o conferem:")
                print(f"   Esperado: {expected_checksum}")
                print(f"   Calculado: {calculated_checksum}")
                return False
                
        except Exception as e:
            print(f"‚ùå Erro durante verifica√ß√£o de integridade: {e}")
            return False
    
    def verify_and_repair_file(self, local_path: str, remote_path: str, max_attempts: int = 3) -> bool:
        """
        Verifica integridade e tenta reparar o arquivo se necess√°rio.
        Faz m√∫ltiplas tentativas de download se a integridade falhar.
        """
        for attempt in range(max_attempts):
            print(f"üîÑ Tentativa {attempt + 1} de {max_attempts}")
            
            if self.verify_file_integrity(local_path, remote_path):
                return True
            
            if attempt < max_attempts - 1:
                print(f"‚ö†Ô∏è Tentativa {attempt + 1} falhou. Tentando novamente...")
                # Remover arquivo corrompido antes da pr√≥xima tentativa
                if os.path.exists(local_path):
                    os.remove(local_path)
                time.sleep(1)  # Pequena pausa antes da pr√≥xima tentativa
        
        print(f"‚ùå Falha na verifica√ß√£o ap√≥s {max_attempts} tentativas")
        return False

    def _close_connections(self):
        """Fecha todas as conex√µes"""
        for channel, _ in self.storage_connections.values():
            channel.close()
        self.storage_connections.clear()
        
        if self.metadata_client:
            self.metadata_client.close()
    
    def _download_chunk_with_fallback(self, arquivo_nome: str, chunk_numero: int, 
                                    chunk_info: fs_pb2.ChunkLocation) -> Optional[bytes]:
        """Download de chunk com fallback para r√©plicas em caso de falha"""
        # Tentar n√≥ prim√°rio primeiro
        primary_node = self.metadata_client.get_node_for_operation("download", arquivo_nome)
        if not primary_node:
            print(f"‚ö†Ô∏è Nenhum n√≥ prim√°rio dispon√≠vel para o arquivo {arquivo_nome}")
            return None
        
        stub = self._get_storage_connection(primary_node.no_recomendado)
        
        # Se chegou aqui, tentar r√©plicas
        if self.metadata_client:
            replicas = self.metadata_client.get_available_replicas(arquivo_nome, chunk_numero)
            for replica in replicas:
                try:
                    replica_stub = self._get_storage_connection(replica)
                    if replica_stub:
                        request = fs_pb2.ChunkRequest(
                            arquivo_nome=arquivo_nome,
                            chunk_numero=chunk_numero
                        )
                        response = replica_stub.DownloadChunk(request)
                        
                        if response.sucesso:
                            import hashlib
                            checksum_calculado = hashlib.md5(response.dados).hexdigest()
                            if checksum_calculado == chunk_info.checksum:
                                print(f"‚úÖ Chunk {chunk_numero} obtido da r√©plica {replica.node_id}")
                                return response.dados
                except Exception as e:
                    print(f"‚ö†Ô∏è Falha na r√©plica {replica.node_id}: {e}")
                    continue
        
        print(f"‚ùå N√£o foi poss√≠vel obter chunk {chunk_numero}")
        return None
    
    def listar(self, caminho: str = "/") -> bool:
        """Lista conte√∫do de diret√≥rio"""
        if not self.metadata_client:
            print("‚ùå Servidor de metadados n√£o dispon√≠vel")
            return False
        
        print(f"\nüìÅ Listando conte√∫do de: '{caminho}' (vis√£o global)")
        
        try:
            # Chama a nova fun√ß√£o que consulta o metadata server
            nomes_arquivos = self.metadata_client.list_files(caminho)
        
            if nomes_arquivos is not None:
                if not nomes_arquivos:
                    print("  (Diret√≥rio vazio)")
                else:
                    for item in nomes_arquivos:
                        print("  üìÑ", item)
                return True
            else:
                # A mensagem de erro j√° foi impressa por list_files_globally
                return False
        except Exception as e:
            print(f"‚ùå Erro na comunica√ß√£o: {e}")
            return False
    
    def upload(self, caminho_local: str, caminho_remoto: str) -> bool:
        """
        Orquestra o upload de um arquivo: registra os metadados, envia os dados
        para um n√≥ de armazenamento e, ao final, marca o arquivo como completo.
        """
        if not os.path.exists(caminho_local):
            print("‚ùå Arquivo local n√£o encontrado.")
            return False
        
        if not self.metadata_client:
            print("‚ùå Servidor de metadados n√£o dispon√≠vel.")
            return False

        try:
            # 1. Preparar os dados do arquivo para o registro
            arquivo_nome = os.path.basename(caminho_remoto)
            with open(caminho_local, "rb") as f:
                dados = f.read()
            
            tamanho_arquivo = len(dados)
            import hashlib
            checksum_arquivo = hashlib.md5(dados).hexdigest()
            # A l√≥gica de quantos chunks ser√£o criados fica no n√≥,
            # mas podemos deixar um placeholder ou um valor padr√£o (1) para o registro inicial.
            total_chunks = 1 

            # 2. REGISTRAR o arquivo no servidor de metadados ANTES do envio
            print(f"INFO: Registrando metadados para '{arquivo_nome}'...")
            sucesso_registro = self.metadata_client.register_file(
                nome_arquivo=arquivo_nome,
                tamanho_total=tamanho_arquivo,
                total_chunks=total_chunks, # O n√≥ atualizar√° isso se dividir o arquivo
                checksum_arquivo=checksum_arquivo
            )
            if not sucesso_registro:
                print("‚ùå Falha ao registrar metadados. Abortando upload.")
                return False
            
            # 3. Obter um n√≥ de armazenamento para enviar o arquivo
            print("INFO: Solicitando n√≥ de armazenamento...")
            node_response = self.metadata_client.get_node_for_operation("upload", arquivo_nome)
            if not (node_response and node_response.sucesso):
                print("‚ùå Nenhum n√≥ de armazenamento dispon√≠vel. Abortando.")
                self.metadata_client.remove_file(arquivo_nome) # Limpa o registro √≥rf√£o
                return False
            
            node_info = node_response.no_recomendado
            
            # 4. ENVIAR o arquivo para o n√≥ de armazenamento
            print(f"INFO: Enviando {tamanho_arquivo} bytes para o n√≥ {node_info.node_id}...")
            stub = self._get_storage_connection(node_info)
            if not stub: return False

            request = fs_pb2.FileUploadRequest(path=caminho_remoto, dados=dados)
            upload_response = stub.Upload(request)
            
            if not upload_response.sucesso:
                print(f"‚ùå Falha no upload para o n√≥: {upload_response.mensagem}")
                return False

            # 5. FINALIZAR: Marcar o upload como completo no servidor de metadados
            # Esta chamada √© importante para o sistema saber que o arquivo est√° pronto para uso.
            print(f"INFO: Finalizando e marcando o arquivo como completo...")
            self.metadata_client.mark_file_complete(arquivo_nome)

            print(f"‚úÖ Upload de '{arquivo_nome}' conclu√≠do com sucesso!")
            return True
        except Exception as e:
            print(f"‚ùå Erro na comunica√ß√£o: {e}")
            return False
    
    def download(self, caminho_remoto: str, caminho_local: str) -> bool:
        """Download de arquivo com suporte autom√°tico a chunks"""
        if not self.metadata_client:
            print("‚ùå Servidor de metadados n√£o dispon√≠vel")
            return False
        
        arquivo_nome = os.path.basename(caminho_remoto)
        
        # Verificar se arquivo existe nos metadados
        file_metadata = self.metadata_client.get_file_metadata(arquivo_nome)
        
        if file_metadata:
            # Arquivo grande com chunks
            print(f"üì• Baixando {arquivo_nome} ({file_metadata.tamanho_total} bytes, {file_metadata.total_chunks} chunks)")
            
            # Obter localiza√ß√£o dos chunks
            chunks_info = self.metadata_client.get_chunk_locations(arquivo_nome)
            if not chunks_info:
                print("‚ùå N√£o foi poss√≠vel obter localiza√ß√£o dos chunks")
                return False
            
            # Download paralelo dos chunks (simplificado - sequencial por enquanto)
            chunks_dados = []
            for chunk_info in sorted(chunks_info, key=lambda x: x.chunk_numero):
                chunk_numero = chunk_info.chunk_numero
                print(f"üì¶ Baixando chunk {chunk_numero + 1}/{len(chunks_info)}")
                
                chunk_data = self._download_chunk_with_fallback(arquivo_nome, chunk_numero, chunk_info)
                if chunk_data is None:
                    print(f"‚ùå Falha ao baixar chunk {chunk_numero}")
                    return False
                
                chunks_dados.append(chunk_data)
            
            # Recombinar chunks
            print("üîß Recombinando chunks...")
            dados_completos = b''.join(chunks_dados)
            
            # Verificar checksum do arquivo completo
            import hashlib
            checksum_calculado = hashlib.md5(dados_completos).hexdigest()
            if checksum_calculado != file_metadata['checksum_arquivo']:
                print("‚ùå Checksum do arquivo n√£o confere")
                return False
            
        else:
            # Tentar download direto (arquivo pequeno ou n√£o encontrado nos metadados)
            node_response = self.metadata_client.get_node_for_operation("download", arquivo_nome)
            if not node_response and node_response.sucesso:
                print("‚ùå Nenhum n√≥ dispon√≠vel para download")
                return False
            
            node_info = node_response.no_recomendado

            stub = self._get_storage_connection(node_info)
            if not stub:
                return False
            
            try:
                request = fs_pb2.CaminhoRequest(path=caminho_remoto)
                response = stub.Download(request)
                
                if not response.sucesso:
                    print("‚ùå Erro no download:", response.mensagem)
                    return False
                
                dados_completos = response.dados
                print(f"üì• Baixando {arquivo_nome} ({len(dados_completos)} bytes) de {node_info['node_id']}")
                
            except Exception as e:
                print(f"‚ùå Erro na comunica√ß√£o: {e}")
                return False
        
        # Salvar arquivo
        try:
            os.makedirs(os.path.dirname(caminho_local), exist_ok=True)
            with open(caminho_local, "wb") as f:
                f.write(dados_completos)
            print(f"‚úÖ Arquivo salvo em: {caminho_local}")
            return True
        except Exception as e:
            print(f"‚ùå Erro ao salvar arquivo: {e}")
            return False
    
    def deletar(self, caminho_remoto: str) -> bool:
        """Deleta arquivo do sistema"""
        if not self.metadata_client:
            print("‚ùå Servidor de metadados n√£o dispon√≠vel")
            return False
        
        arquivo_nome = os.path.basename(caminho_remoto)
        
        # Remover do servidor de metadados (isso remove arquivo e chunks)
        sucesso = self.metadata_client.remove_file(arquivo_nome)
        if sucesso:
            print(f"‚úÖ Arquivo {arquivo_nome} removido do sistema")
            return True
        else:
            print(f"‚ùå Erro ao remover arquivo {arquivo_nome}")
            return False
    
    def copiar(self, origem: str, destino: str) -> bool:
        """Copia arquivo entre locais remotos"""
        if not self.metadata_client:
            print("‚ùå Servidor de metadados n√£o dispon√≠vel")
            return False
        
        # Obter n√≥ para opera√ß√£o de c√≥pia
        node_response = self.metadata_client.get_node_for_operation("copy", origem)
        if not node_response:
            print("‚ùå Nenhum n√≥ dispon√≠vel para c√≥pia")
            return False
        
        node_info = node_response.no_recomendado

        stub = self._get_storage_connection(node_info)
        if not stub:
            return False
        
        try:
            request = fs_pb2.CopyRequest(origem=origem, destino=destino)
            response = stub.Copiar(request)
            
            if response.sucesso:
                print("‚úÖ", response.mensagem)
                return True
            else:
                print("‚ùå Erro na c√≥pia:", response.mensagem)
                return False
        except Exception as e:
            print(f"‚ùå Erro na comunica√ß√£o: {e}")
            return False
    
    def status_sistema(self) -> bool:
        """Exibe status do sistema"""
        if not self.metadata_client:
            print("‚ùå Servidor de metadados n√£o dispon√≠vel")
            return False
        
        try:
            status = self.metadata_client.get_system_status()
            if not status:
                print("‚ùå Erro ao obter status do sistema")
                return False
            
            print("\nüìä Status do Sistema BigFS-v2")
            print("=" * 40)
            print(f"üñ•Ô∏è  N√≥s ativos: {status['nos_ativos']}/{status['total_nos']}")
            print(f"üìÅ Arquivos: {status['total_arquivos']}")
            print(f"üì¶ Chunks: {status['total_chunks']}")
            print(f"üíæ Storage: {status['storage_usado']:,} / {status['storage_total']:,} bytes")
            
            if 'detalhes_nos' in status:
                print("\nüñ•Ô∏è  Detalhes dos N√≥s:")
                for node in status['detalhes_nos']:
                    storage_percent = (node['storage_usado'] / node['capacidade_storage'] * 100) if node['capacidade_storage'] > 0 else 0
                    print(f"  ‚Ä¢ {node['node_id']}: {node['status']} - {storage_percent:.1f}% usado")
            
            if 'estatisticas' in status:
                stats = status['estatisticas']
                print("\nüìà Estat√≠sticas:")
                print(f"  ‚Ä¢ Uploads: {stats['operacoes_upload_total']}")
                print(f"  ‚Ä¢ Downloads: {stats['operacoes_download_total']}")
                print(f"  ‚Ä¢ Deletes: {stats['operacoes_delete_total']}")
                print(f"  ‚Ä¢ Falhas detectadas: {stats['falhas_detectadas']}")
                print(f"  ‚Ä¢ Replica√ß√µes: {stats['replicacoes_realizadas']}")
            
            return True
        except Exception as e:
            print(f"‚ùå Erro ao obter status: {e}")
            return False

    def verificar_integridade_arquivo(self, nome_arquivo: str) -> bool:
            """
            Orquestra a verifica√ß√£o de integridade de todos os chunks de um arquivo,
            comandando cada n√≥ de armazenamento a se autoverificar.
            """
            if not self.metadata_client:
                print("‚ùå Servidor de metadados n√£o dispon√≠vel.")
                return False

            print(f"\nüîç Iniciando verifica√ß√£o de integridade para o arquivo: '{nome_arquivo}'...")

            # 1. Obter a lista de chunks e suas localiza√ß√µes do servidor de metadados.
            chunks_info = self.metadata_client.get_chunk_locations(nome_arquivo)
            if not chunks_info:
                print(f"‚ùå Arquivo n√£o encontrado ou n√£o possui chunks registrados.")
                return False

            total_chunks = len(chunks_info)
            chunks_ok = 0
            erros = []

            # 2. Para cada chunk, pe√ßa ao n√≥ prim√°rio para verificar sua integridade.
            for chunk in sorted(chunks_info, key=lambda c: c.chunk_numero):
                node_id = chunk.no_primario
                chunk_numero = chunk.chunk_numero
                
                print(f"  - Verificando Chunk #{chunk_numero} no N√≥ '{node_id}'...", end=" ")
                
                # Obter as informa√ß√µes de conex√£o do n√≥.
                node_response = self.metadata_client.get_node_info(node_id)
                if not (node_response and node_response.sucesso):
                    print(f"ERRO: N√£o foi poss√≠vel obter informa√ß√µes do n√≥ {node_id}.")
                    erros.append(f"Chunk #{chunk_numero}: Falha ao contatar n√≥ {node_id}.")
                    continue

                node_info = node_response.node_info
                stub = self._get_storage_connection(node_info)
                if not stub:
                    print(f"ERRO: Falha ao conectar ao n√≥ {node_id}.")
                    erros.append(f"Chunk #{chunk_numero}: Falha de conex√£o com {node_id}.")
                    continue

                try:
                    # 3. Enviar o comando de verifica√ß√£o para o n√≥.
                    request = fs_pb2.IntegrityRequest(arquivo_nome=nome_arquivo, chunk_numero=chunk_numero)
                    response = stub.VerificarIntegridade(request)
                    
                    if response.sucesso and response.integridade_ok:
                        print("OK")
                        chunks_ok += 1
                    else:
                        print(f"FALHA ({response.mensagem})")
                        erros.append(f"Chunk #{chunk_numero}: {response.mensagem}")
                except Exception as e:
                    print(f"ERRO DE COMUNICA√á√ÉO ({e})")
                    erros.append(f"Chunk #{chunk_numero}: Erro de comunica√ß√£o com o n√≥.")

            # 4. Apresentar o resultado final.
            print("\n--- Relat√≥rio de Integridade ---")
            print(f"Arquivo: {nome_arquivo}")
            print(f"Chunks Verificados: {total_chunks}")
            print(f"Chunks √çntegros: {chunks_ok}")
            print(f"Status Final: {'‚úÖ SUCESSO' if chunks_ok == total_chunks else '‚ùå FALHA'}")

            if erros:
                print("\nDetalhes dos Erros:")
                for erro in erros:
                    print(f"  - {erro}")
            
            return chunks_ok == total_chunks

def exibir_menu():
    print("\n" + "="*50)
    print("üóÇÔ∏è  BigFS-v2 Client - Sistema de Arquivos Distribu√≠do")
    print("="*50)
    print("1. üìã Listar arquivos (ls)")
    print("2. üóëÔ∏è Deletar arquivo")
    print("3. üì§ Upload de arquivo")
    print("4. üì• Download de arquivo")
    print("5. üìã Copiar arquivo remoto")
    print("6. üìä Status do sistema")
    print("7. üîç Verificar integridade de arquivo")
    print("8. üö™ Sair")
    print("="*50)

def main():
    print("üöÄ Iniciando BigFS-v2 Client...")
    
    # Permitir configura√ß√£o do servidor de metadados
    metadata_server = input("Servidor de metadados (Enter para localhost:50052): ").strip()
    if not metadata_server:
        metadata_server = "localhost:50052"
    
    client = BigFSClient(metadata_server)
    
    try:
        while True:
            exibir_menu()
            escolha = input("Escolha uma op√ß√£o: ").strip()
            
            if escolha == "1":
                caminho = input("Digite o caminho remoto (Enter para raiz): ").strip()
                client.listar(caminho)
                
            elif escolha == "2":
                caminho = input("Digite o caminho do arquivo a ser deletado: ").strip()
                if caminho:
                    client.deletar(caminho)
                else:
                    print("‚ùå Caminho n√£o pode estar vazio")
                    
            elif escolha == "3":
                caminho_local = input("Digite o caminho do arquivo local: ").strip()
                caminho_remoto = input("Digite o caminho remoto de destino: ").strip()
                if caminho_local and caminho_remoto:
                    client.upload(caminho_local, caminho_remoto)
                else:
                    print("‚ùå Caminhos n√£o podem estar vazios")
                    
            elif escolha == "4":
                caminho_remoto = input("Digite o caminho remoto do arquivo: ").strip()
                caminho_local = input("Digite o caminho local para salvar: ").strip()
                if caminho_remoto and caminho_local:
                    client.download(caminho_remoto, caminho_local)
                else:
                    print("‚ùå Caminhos n√£o podem estar vazios")
                    
            elif escolha == "5":
                origem = input("Digite o caminho remoto de origem: ").strip()
                destino = input("Digite o caminho remoto de destino: ").strip()
                if origem and destino:
                    client.copiar(origem, destino)
                else:
                    print("‚ùå Caminhos n√£o podem estar vazios")
                    
            elif escolha == "6":
                client.status_sistema()
                
            elif escolha == "7": # Nova op√ß√£o
                nome_arquivo = input("Digite o nome do arquivo a ser verificado: ").strip()
                if nome_arquivo:
                    client.verificar_integridade_arquivo(nome_arquivo)

            elif escolha == "8": # Ajustar n√∫mero
                print("üëã Encerrando cliente...")
                break
                
            else:
                print("‚ùå Op√ß√£o inv√°lida. Tente novamente.")
                
            input("\nPressione Enter para continuar...")
            
    except KeyboardInterrupt:
        print("\nüëã Cliente interrompido pelo usu√°rio")
    finally:
        client._close_connections()
        print("üîå Conex√µes fechadas")

if __name__ == "__main__":
    main()

